network/src/network.rs

```diff
@@ -315,39 +297,59 @@ impl Network {
         peers_registry.get(peer_id).map(|peer| {
             let protocol_version = peer.protocol_version(protocol_id);
             SessionInfo {
-                peer: PeerInfo {
-                    peer_id: peer_id.to_owned(),
-                    session_type: peer.session.session_type,
-                    last_ping_time: peer.last_ping_time,
-                    connected_addr: peer.connected_addr.clone(),
-                    identify_info: peer.identify_info.clone(),
-                },
+                peer: peer.clone(),
                 protocol_version,
             }
         })
     }
 
-    pub fn dial_addr(&self, addr: Multiaddr) {
-        if let Err(err) = self.p2p_control.write().dial(addr, DialProtocol::All) {
-            error!(target: "network", "failed to dial: {}", err);
-        }
-    }
-
-    pub fn dial(&self, expected_peer_id: &PeerId, mut addr: Multiaddr) {
+    fn inner_dial(
+        &self,
+        expected_peer_id: &PeerId,
+        mut addr: Multiaddr,
+        dial_protocol: DialProtocol,
+    ) -> Result<(), Error> {
         if expected_peer_id == self.local_peer_id() {
             debug!(target: "network", "ignore dial to self");
-            return;
+            return Ok(());
         }
         debug!(target: "network", "dial to peer {:?} address {:?}", expected_peer_id, addr);
-        match Multihash::from_bytes(expected_peer_id.as_bytes().to_vec()) {
-            Ok(peer_id_hash) => {
-                addr.append(multiaddr::Protocol::P2p(peer_id_hash));
-                self.dial_addr(addr);
-            }
-            Err(err) => {
-                error!(target: "network", "failed to convert peer_id to addr: {}", err);
-            }
-        }
+        let p2p_phase =
+            Multihash::from_bytes(expected_peer_id.as_bytes().to_vec()).map_err(|_| {
+                Error::P2P(format!("peer_id is invalid {:?}", expected_peer_id).to_string())
+            })?;
+        addr.append(multiaddr::Protocol::P2p(p2p_phase));
+        self.p2p_control
+            .write()
+            .dial(addr, dial_protocol)
+            .map_err(|_| Error::P2P("dial error".to_string()))
+    }
```

Comment:
Assume I got a peer address, and it's not include a `/p2p` section, than I can't dial it?

Reply:
And very likely one will update his private key, than `/p2p` section is invalid, how to deal with it?

Reply:
CKB network security and scoring base on `peer_id` and `address`(RFC0007), we must enforce connection encrypt.

Reply:
If a peer updated with a new key, In the current implementation, the address(without p2p phase) will consider as invalid, until node rediscovery the new address(with new p2p phase).

Reply:
`connection encrypt` is not require peer_id, you just dial, `secio` will know target peer's peer_id.

Reply:
And, `/p2p` section required in `bootnodes` is not user friendly.

Reply:
Of cause we need `/p2p/` phase to make sure bootnodes' ips not controlled be bad crackers. Pls see RFC0007 first.

Reply:
Yes, `bootnode` must require peer_id, but RFC0007 is not show how peer_id is contribute to the security.

As a normal user, he may change his private key from time to time, but he always use the same address. I think it's a OK behavior, but will he be banned?

Reply:
He **will not be banned**.

The reason we need to deny the old address is as same as the `bootnode` security, a user with good behaviors may earn more scores in his peers' side, once the private key changed, we can't prove the peer's IP is controlled by the same good people or a bad cracker, we must reset the trust(score) to this address.

A normal user may change his key once or twice, but not every time he start the node.

By the way, even the user do it, node outbound connections work as well; only inbound peers will not trust the old score, these peers consider the node as a new node.

There's nothing about * user-friendly*; there's a high probability the user can't feel different after reset key.